File: 3_Examples/observe.py
import cudaq
import numpy as np
from cudaq import spin

qubit_count = 2


@cudaq.kernel
def kernel(qubit_count: int):
    qvector = cudaq.qvector(qubit_count)

    # 2-qubit GHZ state.
    h(qvector[0])
    for i in range(1, qubit_count):
        x.ctrl(qvector[0], qvector[i])

    # If we dont specify measurements, all qubits are measured in
    # the Z-basis by default or we can manually specify it also
    # mz(qvector)


# Define a Hamiltonian in terms of Pauli Spin operators.
hamiltonian = spin.z(0) + spin.y(1) + spin.x(0) * spin.z(0)

# Compute the expectation value given the state prepared by the kernel.
result = cudaq.observe(kernel, hamiltonian, qubit_count).expectation()

print('<H> =', result)

File: 3_Examples/observe.cpp
#include <cudaq.h>
#include <cudaq/spin_op.h>
#include <iostream>

__qpu__ void kernel(int qubit_count)
{
    auto qvector = cudaq::qvector(qubit_count);

    cudaq::h(qvector[0]);

    for (int i : cudaq::range(1, qubit_count))
    {
        cudaq::x<cudaq::ctrl>(qvector[0], qvector[i]);
    }
}

int main(int argc, char const *argv[])
{
    int qubit_count = 2;
    namespace spin = cudaq::spin;
    auto hamiltonian = spin::z(0) + spin::y(1) + spin::x(0) * spin::z(0);

    auto result = cudaq::observe(kernel, hamiltonian, qubit_count).expectation();

    std::cout << '<H> =' << result << std::endl;

    return 0;
}
